dlt/extract/concurrency.py

```diff
@@ -0,0 +1,239 @@
+        """Submit an item to the pool.
+
+        Args:
+            pipe_item: The item to submit.
+            block: If True, block until there's a free slot in the pool.
+
+        Returns:
+            The future if the item was successfully submitted, otherwise None.
+        """
+
+        # Sanity check, negative free slots means there's a bug somewhere
+        assert self.free_slots >= 0, "Worker pool has negative free slots, this should never happen"
+
+        if self.free_slots == 0:
+            if block:
+                # Wait until some future is completed to ensure there's a free slot
+                # Note: This is probably not thread safe. If ever multiple threads will be submitting
+                # jobs to the pool, we ned to change this whole method to be inside a `threading.Lock`
+                self.wait_for_free_slot()
+            else:
+                return None
+
+        future: Optional[TItemFuture] = None
+
+        # submit to thread pool or async pool
+        item = pipe_item.item
+        if isinstance(item, Awaitable):
+            future = asyncio.run_coroutine_threadsafe(item, self._ensure_async_pool())
+        elif callable(item):
+            future = self._ensure_thread_pool().submit(item)
+        else:
+            raise ValueError(f"Unsupported item type: {type(item)}")
+
+        # Future is not removed from self.futures until it's been consumed by the
+        # pipe iterator. But we always want to vacate a slot so new jobs can be submitted
+        future.add_done_callback(self._vacate_slot)
+        self.used_slots += 1
+
+        self.futures[future] = FuturePipeItem(
+            future, pipe_item.step, pipe_item.pipe, pipe_item.meta
+        )
+        return future
+
+    def poll(self) -> None:
+        sleep(self.poll_interval)
+
+    def _resolve_future(self, future: TItemFuture) -> Optional[ResolvablePipeItem]:
+        future, step, pipe, meta = self.futures.pop(future)
+
+        if ex := future.exception():
+            if isinstance(ex, StopAsyncIteration):
+                return None
+            # Raise if any future fails
+            if isinstance(
+                ex, (PipelineException, ExtractorException, DltSourceException, PipeException)
+            ):
+                raise ex
+            raise ResourceExtractionError(pipe.name, future, str(ex), "future") from ex
+
+        item = future.result()
+
+        if item is None:
+            return None
+        elif isinstance(item, DataItemWithMeta):
+            return ResolvablePipeItem(item.data, step, pipe, item.meta)
+        else:
+            return ResolvablePipeItem(item, step, pipe, meta)
+
+    def resolve_next_future(self) -> Optional[ResolvablePipeItem]:
+        """Wait for the next future to be done. Returns None if no futures done."""
+        if not self.futures:
+            return None
+
+        if (item := self.resolve_next_future_no_wait()) is not None:
```

Comment:
It's to guarantee ordered results from each resource. I.e. if there are 2 or more futures from the same pipe already done when this is called, we get the results in FIFO order, since the `self._futures` dict is in insertion order but `as_completed` yields in undefined order.